### ISSUE TITLE
How to add dynamic sql based filters to a cube query for a jsonb postgresql column?

### ISSUE THREAD

**Issue by batflarrow** [cube-js/cube#7009]
**Problem**

I have a table that contains a json column. The column will have a structure something like the following. field_id can vary for different accounts and mostly change at runtime, so I cannot define a dimension in the cube. I want to run queries
```javascript
{
  "field_id" :"value",
   "field_id": ["field_value_1","field_value_2"]
}
```

I want to run queries like the following
```sql
SELECT * FROM my_table
WHERE json_columnn->>'field_id' = queried_value;
```
```sql
SELECT *
FROM my_table
WHERE (json_columnn->'field_id') ?| array['queried_value_1', 'queried_value_2','queried_value_3']
```

Is there a simple way I can achieve this through cube? Is there a way I can specify a filter purely through sql when sending a query? I couldn't find this anywhere in the docs.

**Comment by paveltiunov** [comment]
@batflarrow Right way is to dynamically generate the data model. Please see https://cube.dev/docs/guides/recipes/data-modeling/entity-attribute-value.

**Comment by batflarrow** [comment]
> @batflarrow Right way is to dynamically generate the data model. Please see https://cube.dev/docs/guides/recipes/data-modeling/entity-attribute-value.

Yeah but can we dynamically generate this model for different queries to cube? From what I could understand the data models can only be generated at the time of initialization of the cubejs service, and not for each query.

I can't seem to find a way to dynamically generate a model for each query. I can't generate a model at the time of intiializing the service, because I won't have the information related to all the possible field_ids of the json column at the time of the initialization of the cube service, but I'll have this information at the time of making the queries.

Also even after that, will I be able to add arbitrary SQL filters for the fields that I've added like the ones I've mentioned in the orignal question?

**Comment by igorlukanin** [comment]
@batflarrow Please check this section of the recipe: https://cube.dev/docs/guides/recipes/data-modeling/entity-attribute-value#dynamic-attributes

>  I can't generate a model at the time of intiializing the service, because I won't have the information related to all the possible field_ids of the json column at the time of the initialization of the cube service

That section specifically shows how that can be done.

**Comment by batflarrow** [comment]
> @batflarrow Please check this section of the recipe: https://cube.dev/docs/guides/recipes/data-modeling/entity-attribute-value#dynamic-attributes
> 
> > I can't generate a model at the time of intiializing the service, because I won't have the information related to all the possible field_ids of the json column at the time of the initialization of the cube service
> 
> That section specifically shows how that can be done.

Yeah, I had gone through this. From what I understood, this approach assumes that the statuses are not going to change during the runtime of the cubejs service, but in my case `fieldIds` can keep changing. And they'll be different for different accounts.
Can I somehow invoke this asyncModule on API calls to my backend server?  If yes then I can fetch the `fieldIds` for a particular account and then create a cube schema at runtime for each API call and eventually run a query using that cube.

**Comment by vasishtt** [comment]
Were you able to find a solution on this @batflarrow ?

**Comment by batflarrow** [comment]
> Were you able to find a solution on this @batflarrow ?

Unfortunately, despite my efforts to implement the suggestions from my previous comment, I've had to acknowledge that Cubejs is designed primarily for static schemas. It appears there's no straightforward solution for managing dynamic schemas that evolve during runtime and vary among different users of our application, similar to the ones we have. Despite our best attempts, we were unable to find a direct workaround.

One potential workaround we've devised involves making our dynamic schema somewhat static. Essentially, we've proposed assuming that certain field_ids of specific types will consistently store particular kinds of data. If we can hardcode this relationship into Cubejs Schemas, theoretically, we could achieve what we want. However, we haven't yet implemented this solution in production.

**Comment by igorlukanin** [comment]
@batflarrow I understand that it would require a significant effort but here's what you can do:
* inspect your data for potential filter values with an external service
* generate your data model using that external service
* then use [`repository_factory`](https://cube.dev/docs/reference/configuration/config#repository_factory) and [`schema_version`](https://cube.dev/docs/reference/configuration/config#schema_version) to re-load that data model

**Comment by batflarrow** [comment]
> @batflarrow I understand that it would require a significant effort but here's what you can do:
> 
> * inspect your data for potential filter values with an external service
> * generate your data model using that external service
> * then use [`repository_factory`](https://cube.dev/docs/reference/configuration/config#repository_factory) and [`schema_version`](https://cube.dev/docs/reference/configuration/config#schema_version) to re-load that data model

@igorlukanin I'll look into this. But wanted to check if it's possible to implement this for let's say every API call the the UI makes to my server to fetch the data. And can the data model be different for pretty much each API call?

**Comment by igorlukanin** [comment]
> can the data model be different for every API call?

Well, this might be possible to some extent but this is definitely not a best practice; you would have a huge performance penalty for every query you make.

The right approach is to identify "groups of API calls" that would reuse the same data model. Each group would be a "tenant" in Cube's terms. And for every tenant you can have its own data model.

**Comment by balmychan** [comment]
In my product, schema definitions are dynamically generated by retrieving schema information from a dedicated API.

Below are the assumptions:

- For example, there is a table called organization with a 1..n relationship to item, and the schema definition for custom fields is stored in the organization's itemCustomFields.
- The values of the custom fields are stored in item.values.
- An API is provided to retrieve the custom field definitions via the endpoint `/api/cube/custom-schema?organizationId=${organizationId}`.
 - Access to this API is secured by using a dedicated API key for Cube to make the request.
- Custom fields can have types such as Text, Number, Money, Date, and Datetime.

Below is the sample code:

item.js
```item.js
import { getCustomSchema } from './utils';

asyncModule(async () => {
  const { customDimensions, customMeasures, context } =
    await getCustomSchema('item');

  cube(`item`, {
    sql: `SELECT * FROM item WHERE organizationId = '${context.organizationId}'`,
    data_source: `default`,
    title: 'item',

    joins: {
      organization: {
        sql: `${CUBE.organizationId} = ${organization.id}`,
        relationship: `many_to_one`,
      },
    },

    dimensions: {
      id: {
        sql: `${CUBE}.id`,
        type: `string`,
        title: 'Item ID',
        primary_key: true,
        public: true,
      },

      organizationId: {
        sql: `${CUBE}.organizationId`,
        type: `string`,
        title: `Organization ID`,
        public: false,
      },

      name: {
        sql: `${CUBE}.name`,
        type: `string`,
        title: `Item Name`,
      },

      ...customDimensions,
    },

    measures: {
      count: {
        type: `count`,
        title: `item count`,
      },

      ...customMeasures,
    },

    pre_aggregations: {
      // Pre-aggregation definitions go here.
      // Learn more in the documentation: https://cube.dev/docs/caching/pre-aggregations/getting-started
    },
  });
});
```

utils.js
```utils.js
// eslint-disable-next-line @typescript-eslint/no-var-requires
const fetch = require('node-fetch');
import { apiUrl, apiKey } from '../env';

let promises = {};

export const getCustomSchema = async (type) => {
  const { securityContext } = COMPILE_CONTEXT;

  const organizationId = securityContext?.organizationId ?? 'unknown';
  const schemaVersion = securityContext?.schemaVersion ?? 'unknown';

  const promiseKey = `${organizationId}_${schemaVersion}`;

  if (!promises[promiseKey]) {
    promises[promiseKey] = new Promise((resolve) => {
      fetch(
        `${apiUrl}/api/cube/custom-schema?organizationId=${organizationId}`,
        {
          headers: {
            'X-CUBE-API-Key': apiKey,
          },
        }
      ).then((v) => v.json().then((v2) => resolve(v2)));
    });
  }

  const res = await promises[promiseKey];

  const fields = res[`${type}CustomFields`] ?? [];

  const typeMap = {
    ['Number']: 'number',
    ['Money']: 'number',
    ['Date']: 'time',
    ['DateTime']: 'time',
  };

  return {
    customDimensions: fields.reduce(
      (obj, field) => ({
        ...obj,
        [`values_${field.code}`]: {
          sql: () => `\`values.${field.code}\``,
          type: typeMap[field.type] || 'string',
          title: field.name,
        },
      }),
      {}
    ),
    customMeasures: fields
      .filter((v) => v.type === 'Number' || v.type === 'Money')
      .reduce(
        (obj, field) => ({
          ...obj,
          [`sum_values_${field.code)`]: {
            sql: () => `\`values.${field.code}\``,
            type: `sum`,
            title: `${field.name} Sum`,
            format: 'currency',
          },
        }),
        {}
      ),
    context: {
      organizationId,
    },
  };
};
```

**Comment by ZmeiGorynych** [comment]
It's been a while since the OP, but if anybody with a similar challenge is reading this, we also felt constrained by the not-quite-dynamic nature of Cube, and ended up open sourcing a more dynamic replacement: https://motley.ai/blog-posts/make-it-dynamic (also contains a link to the repo)
